configuration_broker/config_broker.cc

```diff
@@ -0,0 +1,220 @@
+		return nullptr;
+	}
+
+	Debug::log("Unsealed id: {} kind: {} size:{} item: {}",
+	           token->id,
+	           token->kind,
+	           token->maxSize,
+	           token->ConfigId);
+
+	if (token->id == 0)
+	{
+		// Assign an ID so we can track the callbacks added
+		// from this capability
+		static uint16_t nextId = 1;
+		token->id              = nextId++;
+	}
+
+	return token;
+}
+
+//
+// Find a Config by name.  If it doesn't already exist
+// create one.
+//
+NamedConfigItem *find_or_create_config(const char *name)
+{
+	for (auto &c : configData)
+	{
+		if (strcmp(c->name, name) == 0)
+		{
+			return c;
+		}
+	}
+
+	// Allocate a Config object
+	NamedConfigItem *c =
+	  static_cast<NamedConfigItem *>(malloc(sizeof(NamedConfigItem)));
+
+	// Save the name as a read only capability
+	auto nameBuffer = static_cast<char *>(malloc(strlen(name)));
+	strncpy(nameBuffer, name, strlen(name));
+	CHERI::Capability roName{nameBuffer};
+	roName.permissions() &=
+	  roName.permissions().without(CHERI::Permission::Store);
+	c->name         = roName;
+	c->item.version = 0;
+	c->item.data    = nullptr;
+
+	// Add it to the vector
+	configData.push_back(c);
+
+	return c;
+};
+
+//
+// Set a new value for the configuration item described by
+// the capability.
+//
+int __cheri_compartment("config_broker")
+  set_config(SObj sealedCap, void *data, size_t size)
+{
+	Debug::log("thread {} Set config called with {} {} {}",
+	           thread_id_get(),
+	           sealedCap,
+	           data,
+	           size);
+
+	ConfigToken *token = config_capability_unseal(sealedCap);
+	if (token == nullptr)
+	{
+		Debug::log("Invalid capability: {}", sealedCap);
+		return -1;
+	}
+
+	// Check we have a WriteToken
+	if (token->kind != WriteToken)
+	{
+		Debug::log(
+		  "Not a write capability for {}: {}", token->ConfigId, sealedCap);
+		return -1;
+	}
+
+	// Check the size and data are consistent with the token
+	// and each other.
+	if (size > token->maxSize)
+	{
+		Debug::log("invalid size {} for capability: {}", size, sealedCap);
+		return -1;
+	}
+
+	if (size > static_cast<size_t>(Capability{data}.bounds()))
+	{
+		Debug::log("size {} > data.bounds() {}", size, data);
+		return -1;
+	}
+
+	// Allocate heap space for the new value
+	void *newData = malloc(size);
+	if (newData == nullptr)
+	{
+		Debug::log("Failed to allocate space for {}", token->ConfigId);
+		return -1;
+	}
+
+	// Even though we've done the obvious checks were paranoid about
+	// the incomming data so do the copy in a separate compartment
+	if (sandbox_copy(data, newData, size) < 0)
+	{
+		Debug::log("Data copy failed from {} to {}", data, newData);
+		free(newData);
+		return -1;
+	};
+
+	// Find or create a config structure
+	NamedConfigItem *c = find_or_create_config(token->ConfigId);
+
+	// Free the old data value.  Any subscribers that received it should
+	// have thier own claim on it if needed
+	if (c->item.data)
+	{
+		free(c->item.data);
+	}
```

Comment:
Callers have to handle the case anyway, but it’s nicer to reduce the probability of hitting the error-handling paths.

Reply:
If the caller de-reference off item.data other than to pass it into a validator, (which they do after they have the notification of a version change) then there's a risk that they're then using un-validated data.  So it feels in this case that its probably better that they have the old version as long as possible ? 

Maybe get_config() should return a capability that can't be stored so that don't have the risk of reading it again, but can only access it via their own claim ?  (can a clam create a storable capability from a non-storable one ?)

Makes me wonder now if the validator should be passed to get-config() as a callback, so its not possible to get a value that hasn't been through a (still subscriber defined) validator.

Reply:
> So it feels in this case that its probably better that they have the old version as long as possible ?

Yes, that was my intent.  Currently, you are freeing the captured pointer, at which point it will likely go away (unless another subscriber has also claimed it), but you have not yet successfully claimed the new one.  The new one may be freed by the broker in between this and your claim, at which point you have nothing.  You can then try to get the new version, but if the new version fails validation (because your broker does not maintain the invariant that it is publishing only validated data), then what do you do?  

> Maybe get_config() should return a capability that can't be stored so that don't have the risk of reading it again, but can only access it via their own claim ? (can a clam create a storable capability from a non-storable one ?)

You want the thing returned from `get_config` to be captured because it's the thing that you cache.

> Makes me wonder now if the validator should be passed to get-config() as a callback, so its not possible to get a value that hasn't been through a (still subscriber defined) validator.

Possibly, though what happens when two things call `get_config` on the same node?  It feels cleaner to have a separate capability that authorises setting the validator, which you do just once, and then have the broker always maintain the invariant that it has nothing is published unless it passes validation, so subscribers just do a claim.  If they don't actually need to capture the newly published node (if they can process it immediately) then they can do a fast claim.

Either way, it's a bad idea to leave a pointer field set to a dangling pointer.